### ISSUE TITLE
CQT Display Offset

### ISSUE THREAD

**Issue by stefan-balke** [librosa/librosa#586]
#### Description

* Compute CQT.
* Display with `librosa.display.specshow()` and `y_axis='cqt_note'`
* Note axis has offset by one semitone.

#### Steps/Code to Reproduce

```Python
import librosa

cur_audio, fs = librosa.load('Systematic_Scale-C-Major_Meinard_middle.wav')
C = librosa.cqt(cur_audio, sr=fs)
librosa.display.specshow(librosa.amplitude_to_db(C, ref=np.max),
                         sr=fs, x_axis='time', y_axis='cqt_note')```
```

#### Expected Results

For my audio example, first note should be a C4.

#### Actual Results

A C4#
![screen shot 2017-06-09 at 15 06 41](https://user-images.githubusercontent.com/7326732/26976758-869bf1a4-4d25-11e7-9ca9-19858575c3c1.png)

I think the problem lies here:
https://github.com/librosa/librosa/blob/master/librosa/display.py#L855

Instead of
`core.cqt_frequencies(n+1, fmin=fmin, bins_per_octave=bins_per_octave)`

we need

`core.cqt_frequencies(n, fmin=fmin, bins_per_octave=bins_per_octave)`

The audio file can be obtained from here: https://github.com/stefan-balke/mpa-exc/blob/master/data/Systematic_Scale-C-Major_Meinard_middle.wav

**Comment by bmcfee** [comment]
I'm not able to reproduce this directly.  Would you mind posting a complete example to generate the figure you posted?  (It looks like you've zoomed in on a region.)  Also, could you post the `librosa.show_versions()` output?

When I try to replicate this locally, everything appears to work properly.  In nbagg mode, the mouseover on the bottom row reports `C1` -- when I zoom in, it shows `C1+16`, which is just an artifact of the ticker placement in absolute coordinates.
```python
y, sr = librosa.load(librosa.util.example_audio_file(), offset=10, duration=2)

C = librosa.cqt(y=y, sr=sr)

plt.figure()
librosa.display.specshow(librosa.amplitude_to_db(C, ref=np.max),
                        sr=sr, x_axis='time', y_axis='cqt_note')
plt.xlim([0, 2])
plt.ylim([30, 60])
```
![image](https://user-images.githubusercontent.com/1190540/26977749-faf2b940-4cf6-11e7-829f-225d4f621c35.png)

**Comment by stefan-balke** [comment]
Here you go.

Code and Audio: [lr_cqt.zip](https://github.com/librosa/librosa/files/1064210/lr_cqt.zip)

Should result in this:
 
![screen shot 2017-06-09 at 15 51 24](https://user-images.githubusercontent.com/7326732/26978346-880c659a-4d2b-11e7-9e68-2618eaea82a5.png)


```
INSTALLED VERSIONS
------------------
python: 3.6.0 |Continuum Analytics, Inc.| (default, Dec 23 2016, 13:19:00) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]

librosa: 0.5.1

audioread: installed, no version number available
numpy: 1.12.1
scipy: 0.19.0
scikit-learn: None
joblib: 0.11
decorator: 4.0.11
six: 1.10.0
resampy: 0.1.5

numpydoc: None
sphinx: None
sphinx_rtd_theme: None
sphinxcontrib-versioning: None
matplotlib: 2.0.2
numba: None
None
```

**Comment by bmcfee** [comment]
For posterity:

```python
print(librosa.show_versions())

cur_audio, fs = librosa.load('Systematic_Scale-C-Major_Meinard_middle.wav')

C = librosa.cqt(cur_audio, sr=fs)
librosa.display.specshow(librosa.amplitude_to_db(C, ref=np.max),
                         sr=fs, x_axis='time', y_axis='cqt_note')
plt.xlim([0.9, 1.1])
plt.ylim([220, 440])
plt.show()
```

So I think what's happening here is that the bin centers are indeed correct, and there's no bug.  When you zoom in to the level where cent deviations are plotted, matplotlib will try to evenly space ticks in the visible region.  In the example you show, the bottom cell gets ticked at `A#3-38` rather than `A3+62` because the tick lands above the half-way point of the bin (by 12 cents, in fact).  This causes it to round up to `A#` with a negative tuning.

Note that this label only applies to the tick not the entire bin.  I agree that it's perhaps a little confusing, but I believe it's the correct behavior if you consider overlaying a pitch contour plot on top of the CQT -- you should see that the ticks are correctly labeled and in the correct position.

**Comment by stefan-balke** [comment]
Okay, my point is that I know that the first note is a C4.

If I run it through the semitone filterbank with same plot settings I get this:

![plot](https://user-images.githubusercontent.com/7326732/26981093-624d424e-4d34-11e7-99b5-638cc6188441.png)

```Python
import numpy as np
import librosa
import librosa.display
import matplotlib.pyplot as plt


# Load audio
cur_audio, fs = librosa.load('Systematic_Scale-C-Major_Meinard_middle.wav')

# Pitch Features (Semitones)
win_len = 4410
hop_size = 2205
f_pitch = librosa.semitone_spectrogram(cur_audio, hop_length=hop_size, win_length=win_len)

#  CENS Features
f_chroma = librosa.feature.chroma_cqt(C=f_pitch, sr=fs, tuning=0, hop_length=hop_size,
                                      fmin=librosa.midi_to_hz(21))

# Plotting
plt.subplot(2, 1, 1)
librosa.display.specshow(librosa.logamplitude(f_pitch**2, ref_power=np.max),
                         y_axis='cqt_note', cmap='gray_r', hop_length=hop_size,
                         x_axis='time', fmin=librosa.midi_to_hz(21))
plt.title('Semitone Spectrogram')
plt.xlim([0.9, 1.1])
plt.ylim([220, 440])

plt.subplot(2, 1, 2)
librosa.display.specshow(f_chroma,
                         y_axis='chroma', cmap='gray_r', hop_length=hop_size,
                         x_axis='time')
plt.title('CENS Features')
plt.xlim([0.9, 1.1])

plt.tight_layout()
plt.savefig('plot.png')
plt.show()
```

Maybe there is something going on in the filterbank but I double-checked it. Also, I think it's strange that "CQT" and Chroma are having a semitone offset.

**Comment by stefan-balke** [comment]
Wait...let me read your post again.

**Comment by stefan-balke** [comment]
Alright, I think I know what you mean.

I'm not concerned about the ticks and how they deviate in cents because they are not plotted in the middle of the bin, I'm unhappy that although the first note is a C4, it lands in the C#4 band in both the CQT (as far as one can tell) and in the filterbank approach. That's why I have the impressions that the axis labels are off by one semitone.

*Regarding the tick mark locations:*
I think this is really misleading and gives the impression of a frequency resolution that is not in the data itself.

**Comment by bmcfee** [comment]
> it lands in the C#4 band in both the CQT (as far as one can tell) and in the filterbank approach.

It's not in the C#4 band.  It's just that the tick that gets placed within the band boundaries happens to be closer to the C# side than the C side, so that's what it uses as a reference.  I see this as an artifact of plotting bin-quantized frequency measurements in absolute coordinates and allowing the ticks to be placed arbitrarily.  Maybe there's a way to lock tick placement to bin centers, but I'm wary of getting too hands-on with the plot decorations because those things get complicated really quickly.



> Regarding the tick mark locations:
> I think this is really misleading and gives the impression of a frequency resolution that is not in the data itself.

I agree with this, but don't have a good idea for improving it.  Maybe someone else does?

**Comment by stefan-balke** [comment]
I see. Is this a behaviour of `pcolormesh()`? Never saw that when I simply used `imshow()` and set my tick labels manually.

**Comment by bmcfee** [comment]
> Is this a behaviour of pcolormesh()?

Sort of -- when we [call](https://github.com/librosa/librosa/blob/master/librosa/display.py#L673-L679) `pcolormesh` we provide the bin boundaries as provided by the `cqt` frequency map.  The `n+1` you previously referred to is there to indicate the upper edge of the top bin.

Now that I look at it closely, maybe we should do it a little differently though.  `pcolormesh` expects bin `i` to span coords `(y[i], y[i+1])`.  Using `cqt_frequencies` directly from `fmin`, this puts the center frequency of a filter `i` at `y[i]` so that its center is actually `y[i] + (y[i+1] - y[i])/2`, which might explain the cent deviations.  We could easily recenter this so that bins are centered at their center frequency, rather than start at the center frequency.

**Comment by bmcfee** [comment]
I took an initial cut at this, adjusting the mesh coordinates to be centered at the CQT frequencies, rathre than left-aligned.  The following example code should then place x's in the (vertical) center of each bin:
```python
C = librosa.amplitude_to_db(librosa.cqt(y=y, sr=sr), ref=np.max)

plt.figure()
librosa.display.specshow(C, y_axis='cqt_note')
plt.plot(librosa.cqt_frequencies(len(C), librosa.note_to_hz('C1')), c='w', marker='x')
plt.xlim([0, 10])
plt.ylim([32, 64])
```
as shown here:
![image](https://user-images.githubusercontent.com/1190540/27973137-ff88f10e-6326-11e7-9e15-68f768182290.png)

This doesn't fix the tick location problem, which as mentioned above, is going to be rather involved and involve modifications to different parts of the display module.

While I'm at it, it's probably worth auditing the other grid layouts (linear, log, mel, tempo) for bin centering consistency as well.

**Comment by bmcfee** [comment]
This should be fixed by merging #597 .  @stefan-balke -- feel free to repon if the issue is not sufficiently addressed.
